bfvmm/src/vmcs/src/bitmap.cpp

```diff
@@ -0,0 +1,50 @@
+#include <vmcs/bitmap.h>
+
+bitmap::bitmap(uint32_t num_bits)
+{
+    m_length = num_bits / 8;
+
+    if (num_bits % 8)
+    {
+        m_length++;
+    }
+
+    m_bitmap = new uint8_t[m_length];
+
+    for (uint32_t i = 0; i < num_bits; i++)
+    {
+        set_bit(i);
+        clear_bit(i);
+    }
+}
+
+uint8_t *bitmap::address()
+{
+    return m_bitmap;
+}
+
+void bitmap::set_bit(uint32_t n)
+{
+    if ((n / 8) > m_length) return;
+
+    m_bitmap[n / 8] |= (1 << (n % 8));
+}
+
+void bitmap::clear_bit(uint32_t n)
+{
+    if ((n / 8) > m_length) return;
+
+    m_bitmap[n / 8] &= ~(1 << (n % 8));
+}
+
+bool bitmap::bit(uint32_t n)
+{
+    if ((n / 8) > m_length) return false;
```

Comment:
There are a couple of issues here. First, we should really try to avoid / in the hypervisor as it "could" cause problems. In this case it would not as you are using a constant.

Are you sure this actually works? Seems to me that if n == 2, you get 2/8 = 0, and if n == 3, you get 3/8 == 0.